File: benchmark/harness/Sources/BenchmarkKit/GenLib.swift
// GenLib.swift — shared helpers for the two deterministic generators, port of
// benchmark/lib/genlib.py adapted to the Swift TTT fixture.
//
// Owns: locating PLUGIN_ROOT / ttt-template, copying the template into a per-side
// workdir (staged or single-shot), running the generated app's Swift Testing
// suite (parsing REAL test_count/pass_fail), counting REAL loc_produced (*.swift),
// and building a PathMetrics from measured values. Never imports BenchmarkLive.
//
// AR contract deltas honored here:
//   - locCount counts *.swift, EXCLUDING .build/ and .swiftpm/ (was *.py).
//   - runAppTests: pass_fail from the exit code ONLY (never parse success text);
//     test_count from the Swift Testing summary regex `Test run with (\d+) test`.
//   - staged copy = one stage per Sources/TicTacToeKit/<subdir> + executable +
//     Tests (was one-per-.py). stage_count is a process-overhead signal; the
//     exact count is NOT a contract (assert >1 WITH, ==1 WITHOUT).

import Foundation

public enum GenLib {
    // MARK: path anchoring

    /// Repo root. Resolved from the running binary's location when possible,
    /// else from an explicit override (tests inject a temp template).
    public static func pluginRoot(from benchmarkDir: String) -> String {
        (benchmarkDir as NSString).deletingLastPathComponent
    }

    // MARK: monotonic timer

    /// Monotonic wall-clock timer. `elapsed` is seconds since `start()`.
    public final class Timer {
        private var t0: UInt64 = 0
        public init() {}
        public func start() { t0 = DispatchTime.now().uptimeNanoseconds }
        public var elapsed: Double {
            Double(DispatchTime.now().uptimeNanoseconds - t0) / 1_000_000_000
        }
        /// Time a closure, returning (result, elapsedSeconds).
        public static func measure<T>(_ body: () throws -> T) rethrows -> (T, Double) {
            let t = Timer(); t.start()
            let r = try body()
            return (r, t.elapsed)
        }
    }

    // MARK: template copy

    static let ignoredDirNames: Set<String> = [".build", ".swiftpm", "__pycache__"]

    private static func copyTree(from src: String, to dest: String) throws {
        let fm = FileManager.default
        try fm.createDirectory(atPath: dest, withIntermediateDirectories: true)
        for name in (try? fm.contentsOfDirectory(atPath: src)) ?? [] {
            if ignoredDirNames.contains(name) || name.hasSuffix(".pyc") { continue }
            let s = (src as NSString).appendingPathComponent(name)
            let d = (dest as NSString).appendingPathComponent(name)
            var isDir: ObjCBool = false
            fm.fileExists(atPath: s, isDirectory: &isDir)
            if isDir.boolValue {
                try copyTree(from: s, to: d)
            } else {
                try? fm.removeItem(atPath: d)
                try fm.copyItem(atPath: s, toPath: d)
            }
        }
    }

    /// WITHOUT path: copy the whole template in ONE operation. Returns op count (1).
    @discardableResult
    public static func copyTemplateSingleShot(templateDir: String, dest: String) throws -> Int {
        let fm = FileManager.default
        try fm.removeItemIfExists(atPath: dest)
        try copyTree(from: templateDir, to: dest)
        return 1
    }

    /// WITH path: materialize the template as N discrete staged operations.
    /// One stage per Sources/TicTacToeKit/<subdir> + one for the executable +
    /// one for Package.swift + one for Tests. Returns the staged-op count.
    @discardableResult
    public static func copyTemplateStaged(templateDir: String, dest: String) throws -> Int {
        let fm = FileManager.default
        try fm.removeItemIfExists(atPath: dest)
        try fm.createDirectory(atPath: dest, withIntermediateDirectories: true)
        var stages = 0

        // Stage 1: Package.swift (+ any top-level manifest files).
        let manifest = (templateDir as NSString).appendingPathComponent("Package.swift")
        if fm.fileExists(atPath: manifest) {
            try fm.copyItem(atPath: manifest,
                            toPath: (dest as NSString).appendingPathComponent("Package.swift"))
            stages += 1
        }

        // Stage 2..N: one stage per Sources/TicTacToeKit/<subdir>, then the exe.
        let kitSrc = templateDir + "/Sources/TicTacToeKit"
        let kitDst = dest + "/Sources/TicTacToeKit"
        if fm.fileExists(atPath: kitSrc) {
            let subdirs = ((try? fm.contentsOfDirectory(atPath: kitSrc)) ?? []).sorted()
            for sub in subdirs {
                if ignoredDirNames.contains(sub) { continue }
                let s = (kitSrc as NSString).appendingPathComponent(sub)
                let d = (kitDst as NSString).appendingPathComponent(sub)
                var isDir: ObjCBool = false
                fm.fileExists(atPath: s, isDirectory: &isDir)
                if isDir.boolValue {
                    try copyTree(from: s, to: d)
                } else {
                    try fm.createDirectory(atPath: kitDst, withIntermediateDirectories: true)
                    try fm.copyItem(atPath: s, toPath: d)
                }
                stages += 1
            }
        }

        // Executable target stage.
        let exeSrc = templateDir + "/Sources/tictactoe"
        if fm.fileExists(atPath: exeSrc) {
            try copyTree(from: exeSrc, to: dest + "/Sources/tictactoe")
            stages += 1
        }

        // Final stage: Tests dir.
        let testsSrc = (templateDir as NSString).appendingPathComponent("Tests")
        if fm.fileExists(atPath: testsSrc) {
            try copyTree(from: testsSrc, to: (dest as NSString).appendingPathComponent("Tests"))
            stages += 1
        }

        return stages
    }

    // MARK: measurement

    /// REAL loc_produced: non-blank, non-pure-comment lines across all generated
    /// *.swift files (excluding .build/.swiftpm).
    public static func countLOC(appDir: String) -> Int {
        var total = 0
        let fm = FileManager.default
        guard let e = fm.enumerator(atPath: appDir) else { return 0 }
        for case let rel as String in e {
            if rel.contains(".build/") || rel.contains(".swiftpm/") { continue }
            guard rel.hasSuffix(".swift") else { continue }
            let full = (appDir as NSString).appendingPathComponent(rel)
            guard let text = try? String(contentsOfFile: full, encoding: .utf8) else { continue }
            for line in text.split(separator: "\n", omittingEmptySubsequences: false) {
                let s = line.trimmingCharacters(in: .whitespaces)
                if !s.isEmpty && !s.hasPrefix("//") { total += 1 }
            }
        }
        return total
    }

    /// Run the generated app's Swift Testing suite. Returns (test_count, pass_fail).
    /// test_count parsed from the Swift Testing summary line
    /// `Test run with N test(s) ...`; pass_fail from the EXIT CODE only.
    ///
    /// OI-1 (P1): the nested `swift test` materializes a ~150MB `.build/` cache
    /// INSIDE `appDir`. Nothing downstream reads it (loc counts *.swift and skips
    /// `.build/`; metrics reference only `appDir`/Package.swift), so it is the
    /// ephemeral heavy artifact. A `defer` sweeps it on EVERY exit path (normal
    /// return OR a thrown parse error), scoped strictly to `appDir/.build` so the
    /// caller's measured app copy and any `--workdir`/`--record` path survive
    /// (R3). This bounds residue to the app sources, not the build cache — the
    /// two ENOSPC-killed live attempts were this cache accumulating across runs.
    public static func runAppTests(appDir: String) -> (testCount: Int, passFail: String) {
        let buildDir = (appDir as NSString).appendingPathComponent(".build")
        defer { try? FileManager.default.removeItemIfExists(atPath: buildDir) }
        let result = Subprocess.run(
            ["swift", "test", "--package-path", appDir],
            cwd: appDir
        )
        let combined = result.stdout + result.stderr
        let count = parseTestCount(combined)
        let passFail = result.exitCode == 0 ? "pass" : "fail"
        return (count, passFail)
    }

    /// Parse the Swift Testing suite size from a run log. Matches lines like
    /// "Test run with 27 tests" / "✔ Test run with 1 test passed". Falls back to
    /// XCTest-style "Executed N tests" if present. 0 when unparseable.
    public static func parseTestCount(_ log: String) -> Int {
        // Swift Testing: "Test run with N test(s)"
        if let n = firstMatchInt(log, pattern: #"Test run with (\d+) test"#) { return n }
        // XCTest fallback: "Executed N tests"
        if let n = firstMatchInt(log, pattern: #"Executed (\d+) test"#) { return n }
        return 0
    }

    private static func firstMatchInt(_ text: String, pattern: String) -> Int? {
        guard let re = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(text.startIndex..., in: text)
        guard let m = re.firstMatch(in: text, range: range), m.numberOfRanges >= 2,
              let r = Range(m.range(at: 1), in: text) else { return nil }
        return Int(text[r])
    }

    /// Build a PathMetrics from measured values (tokens null in deterministic).
    public static func buildPathMetrics(
        appDir: String, pluginRoot: String, stageCount: Int, estimateComplexityScore: Int,
        costUSD: Double?, wallClockS: Double
    ) -> PathMetrics {
        let (testCount, passFail) = runAppTests(appDir: appDir)
        let loc = countLOC(appDir: appDir)
        // Repo-relative app_path so committed history never leaks an absolute path.
        let appPath = relativePath(appDir, from: pluginRoot)
        return PathMetrics(
            tokens: Tokens(input: nil, output: nil, total: nil),
            costUSD: costUSD,
            wallClockS: (wallClockS * 10000).rounded() / 10000,
            locProduced: loc,
            testCount: testCount,
            coveragePct: 0.0,
            estimateComplexityScore: estimateComplexityScore,
            stageCount: stageCount,
            passFail: passFail,
            appPath: appPath
        )
    }

    static func relativePath(_ path: String, from base: String) -> String {
        let p = (path as NSString).standardizingPath
        let b = (base as NSString).standardizingPath
        if p.hasPrefix(b + "/") { return String(p.dropFirst(b.count + 1)) }
        return p
    }
}

// MARK: - Subprocess helper (shared by GenLib / Generators / bench-*)

public struct Subprocess {
    public var exitCode: Int32
    public var stdout: String
    public var stderr: String

    /// Run `argv` synchronously, capturing stdout/stderr. `input` is fed on stdin.
    @discardableResult
    public static func run(_ argv: [String], cwd: String? = nil, input: String? = nil,
                           env: [String: String]? = nil) -> Subprocess {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = argv
        if let cwd { process.currentDirectoryURL = URL(fileURLWithPath: cwd) }
        if let env { process.environment = env }

        let outPipe = Pipe(), errPipe = Pipe()
        process.standardOutput = outPipe
        process.standardError = errPipe
        let inPipe = Pipe()
        if input != nil {
            process.standardInput = inPipe
        } else {
            // Never inherit the caller's stdin: in nested/headless contexts an
            // inherited tty/pipe stdin can block a child (claude auth status,
            // estimate-calc.py, …) forever. No input => explicit EOF.
            process.standardInput = FileHandle.nullDevice
        }

        // Drain pipes on background queues to avoid deadlock on large output.
        // Lock-protected boxes keep the closure captures Sendable-safe.
        let outBox = DataBox(), errBox = DataBox()
        let group = DispatchGroup()
        let q = DispatchQueue(label: "subprocess.drain", attributes: .concurrent)
        do {
            try process.run()
        } catch {
            return Subprocess(exitCode: 127, stdout: "", stderr: "spawn failed: \(error)")
        }
        if let input {
            inPipe.fileHandleForWriting.write(Data(input.utf8))
            try? inPipe.fileHandleForWriting.close()
        }
        group.enter()
        q.async { outBox.store(outPipe.fileHandleForReading.readDataToEndOfFile()); group.leave() }
        group.enter()
        q.async { errBox.store(errPipe.fileHandleForReading.readDataToEndOfFile()); group.leave() }
        process.waitUntilExit()
        group.wait()
        return Subprocess(
            exitCode: process.terminationStatus,
            stdout: String(decoding: outBox.value, as: UTF8.self),
            stderr: String(decoding: errBox.value, as: UTF8.self)
        )
    }
}

/// Thread-safe byte buffer for pipe draining (avoids captured-var mutation in
/// concurrently-executing code).
private final class DataBox: @unchecked Sendable {
    private let lock = NSLock()
    private var data = Data()
    func store(_ d: Data) { lock.lock(); data = d; lock.unlock() }
    var value: Data { lock.lock(); defer { lock.unlock() }; return data }
}

File: benchmark/harness/benchmarkkit/genlib.py
"""Process execution and app metrics for the two deterministic generators.

Runs the generated app's Swift Testing suite (REAL ``swift test`` — the retained
measurement instrument), counts REAL loc_produced (*.swift), and builds a
PathMetrics from measured values. Template materialization lives in ``treecopy``.
Never imports benchmarklive.

OI-1: the nested ``swift test`` materializes a heavy ``.build/`` inside the app
dir; a ``finally`` sweeps exactly ``<app_dir>/.build`` on every exit path, scoped
so the caller's app copy / --workdir survive (R3).
"""

from __future__ import annotations

import os
import re
import shutil
import subprocess
import time
from dataclasses import dataclass
from typing import Optional

from .metrics import PathMetrics, Tokens


class Timer:
    """Monotonic wall-clock timer; ``elapsed`` is seconds since ``start()``."""

    def __init__(self) -> None:
        self._t0 = 0.0

    def start(self) -> None:
        self._t0 = time.monotonic()

    @property
    def elapsed(self) -> float:
        return time.monotonic() - self._t0


@dataclass
class Subprocess:
    exit_code: int
    stdout: str
    stderr: str

    @staticmethod
    def run(argv: list, cwd: Optional[str] = None, input: Optional[str] = None,
            env: Optional[dict] = None, timeout: Optional[float] = None) -> "Subprocess":
        try:
            # No input → stdin is /dev/null (never inherit the caller's stdin; a
            # nested/headless child can block forever on an inherited tty).
            r = subprocess.run(
                argv,
                cwd=cwd,
                env=env,
                input=input,
                stdin=None if input is not None else subprocess.DEVNULL,
                capture_output=True,
                text=True,
                timeout=timeout,
            )
        except subprocess.TimeoutExpired:
            # rc 124 (shell/coreutils `timeout` convention); never leak the prompt.
            return Subprocess(124, "", f"timed out after {timeout}s")
        except (OSError, ValueError) as exc:
            return Subprocess(127, "", f"spawn failed: {exc}")
        return Subprocess(r.returncode, r.stdout, r.stderr)


def plugin_root(benchmark_dir: str) -> str:
    return os.path.dirname(benchmark_dir)


def count_loc(app_dir: str, exclude_dirs: Optional[set] = None) -> int:
    """REAL loc_produced: non-blank, non-pure-comment lines across generated *.swift.

    ``exclude_dirs`` names (not paths) are pruned at the app_dir root only, so a
    WITH-root count can skip a sibling ``without/`` arm directory.
    """
    skip = {".build", ".swiftpm"} | (exclude_dirs or set())
    total = 0
    for root, dirs, files in os.walk(app_dir):
        if root == app_dir:
            dirs[:] = [d for d in dirs if d not in skip]
        else:
            dirs[:] = [d for d in dirs if d not in (".build", ".swiftpm")]
        for name in files:
            if not name.endswith(".swift"):
                continue
            try:
                with open(os.path.join(root, name), encoding="utf-8") as f:
                    text = f.read()
            except OSError:
                continue
            for line in text.split("\n"):
                s = line.strip()
                if s and not s.startswith("//"):
                    total += 1
    return total


def parse_test_count(log: str) -> int:
    """Swift Testing 'Test run with N test(s)'; XCTest 'Executed N tests' fallback."""
    m = re.search(r"Test run with (\d+) test", log)
    if m:
        return int(m.group(1))
    m = re.search(r"Executed (\d+) test", log)
    if m:
        return int(m.group(1))
    return 0


def run_app_tests(app_dir: str) -> tuple:
    """Run the generated app's suite. Returns (test_count, pass_fail).

    pass_fail from the EXIT CODE only; test_count from the summary regex. OI-1:
    ``<app_dir>/.build`` swept in ``finally`` on every path.
    """
    build_dir = os.path.join(app_dir, ".build")
    try:
        r = Subprocess.run(["swift", "test", "--package-path", app_dir], cwd=app_dir)
        combined = r.stdout + r.stderr
        return parse_test_count(combined), ("pass" if r.exit_code == 0 else "fail")
    finally:
        shutil.rmtree(build_dir, ignore_errors=True)


def relative_path(path: str, base: str) -> str:
    p = os.path.normpath(path)
    b = os.path.normpath(base)
    if p.startswith(b + os.sep):
        return p[len(b) + 1:]
    return p


def build_path_metrics(app_dir: str, plugin_root_dir: str, stage_count: int,
                       estimate_complexity_score: int, cost_usd: Optional[float],
                       wall_clock_s: float, oracle_result: Optional[dict] = None) -> PathMetrics:
    test_count, self_graded = run_app_tests(app_dir)
    loc = count_loc(app_dir)
    # The held-out oracle outranks the arm's own suite wherever it ran.
    pass_fail = self_graded
    if oracle_result is not None:
        pass_fail = "pass" if oracle_result.get("pass_rate") == 1.0 else "fail"
    return PathMetrics(
        tokens=Tokens(input=None, output=None, total=None),
        cost_usd=cost_usd,
        wall_clock_s=round(wall_clock_s * 10000) / 10000,
        loc_produced=loc,
        test_count=test_count,
        coverage_pct=None,  # nothing here measures coverage; a real 0.0 would claim it did
        estimate_complexity_score=estimate_complexity_score,
        stage_count=stage_count,
        pass_fail=pass_fail,
        app_path=relative_path(app_dir, plugin_root_dir),
        oracle=oracle_result,
    )
